### PR TITLE
Fix nullable detection

### LINKED ISSUE #11828: Disabled Nullability tests
https://github.com/aspnet/AspNetCore/pull/11818 disables the following tests due to ingesting the new Roslyn compiler, which changes some nullability metadata:

CanUseNonNullableReferenceType_WithController_OmitData_ValidationErrors
BindProperty_WithNonNullableReferenceType_NoData_ValidationError
BindProperty_WithNonNullableReferenceType_NoData_ValidationError_CustomMessage
BindParameter_WithNonNullableReferenceType_NoData_ValidationError
CreateValidationMetadata_InfersRequiredAttribute_NoNonNullableProperty
IsNonNullable_FindsNonNullableProperty
IsNonNullable_FindsNonNullableParameter

We should look into what's needed to re-enable these.

CC @dougbu @rynowak

### PR DESCRIPTION
**Pull request by rynowak**
Fixes: #11828 and #11813

### CODE REVIEW

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs [hunk 1]
```diff
@@ -443,10 +484,51 @@ internal static bool IsNonNullable(IEnumerable<object> attributes)
...
             }
 
-            return false;
+            isNullable = false;
+            return true; // [Nullable] found but type is not an NNRT
+        }
+
+        internal static bool IsNullableBasedOnContext(Type containingType, MemberInfo member)
+        {
+            var attributes = member?.GetCustomAttributes(inherit: true) ?? Array.Empty<object>();
```
Comment (roji) [inline src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: According to https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-metadata.md, NullableContextAttribute is not inherited.

Comment (roji) [inline src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: I didn't handle NullableContextAttribute at the member level - should do that. I know it's explicitly mentioned in the metadata doc, but any idea what that means and why it's useful?

**Review by roji — APPROVED** [review]
LGTM, two comments:

* Not sure if model processing speed is an issue, but you may want to look at caching context nullability at the type and module level, to avoid doing a lot of reflection over and over again. In EF Core we've seen some ~insane~ prolific models where this could be an issue.
* I can see the comment about not handling generics and NNRT, but we may want to consider identifying pre- and post-conditions such as [MaybeNull] ([fully explained here](https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md)).

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs (hunk 1 above)
Comment (rynowak) [reply src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: We process parameters too, so it's possible for `NCA` to appear on the method declaration.

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs [hunk 2]
```diff
@@ -443,10 +484,51 @@ internal static bool IsNonNullable(IEnumerable<object> attributes)
...
+            var isNullable = AttributesHasNullableContext(attributes);
+            if (isNullable != null)
+            {
+                return isNullable.Value;
+            }
+
+            attributes = containingType.GetCustomAttributes(inherit: false);
```
Comment (pranavkm) [inline src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: The `inherits: false` warrants a commit. We generally don't do this elsewhere.

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs [hunk 3]
```diff
@@ -443,10 +484,51 @@ internal static bool IsNonNullable(IEnumerable<object> attributes)
...
+
+            // If we don't find the attribute on the declaring type then repeat at the module level
+            attributes = containingType.Module.GetCustomAttributes(inherit: false);
+            isNullable = AttributesHasNullableContext(attributes);
+            return isNullable ?? false;
+
+            bool? AttributesHasNullableContext(object[] attributes)
+            {
+                var nullableContextAttribute = attributes
+                    .Where(a => string.Equals(a.GetType().FullName, NullableContextAttributeFullName, StringComparison.Ordinal))
```
Comment (pranavkm) [inline src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: `.FirstOrDefault(...);`

**Review by pranavkm — APPROVED** [review]

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs [hunk 4]
```diff
@@ -350,20 +356,44 @@ public void CreateValidationMetadata(ValidationMetadataProviderContext context)
...
+                        context.Key.ParameterInfo?.Member.ReflectedType, 
+                        context.Key.ParameterInfo.Member, 
+                        context.ParameterAttributes);
+                }
+                else
                 {
-                    AllowEmptyStrings = true,
-                };
-                attributes.Add(requiredAttribute);
+                    throw new InvalidOperationException("Unsupported ModelMetadataKind: " + context.Key.MetadataKind);
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: No, we don't do this in other parts of the product, and we have no plan to actually localize the error messages. We should stop paying a tax for this.

I want this to throw an exception if it happens (it won't).

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs (hunk 2 above)
Comment (rynowak) [reply src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: sure. Will add comments for these - `[Nullable]` and `[NullableContext]` aren't inherited, so there's no value in doing the extra work.

File: src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs (hunk 3 above)
Comment (rynowak) [reply src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs]: sure

**Comment by rynowak** [comment]
Thanks @roji 

I'm not immediately concerned about intermediate caching, the results of this will all be cached as well. 

> I can see the comment about not handling generics and NNRT, but we may want to consider identifying pre- and post-conditions such as [MaybeNull]

Are these types actually defined anywhere? I keep seeing design notes, but I don't see these types in use **anywhere**.

**Comment by roji** [comment]
> Are these types actually defined anywhere? I keep seeing design notes, but I don't see these types in use anywhere.

Take a look in corefx, they're used a lot there. BTW unlike NullableAttribute and NullableContextAttribute, these are actually defined in System.Diagnostics.CodeAnalysis (not sure why the different approach).

Note also https://github.com/dotnet/corefx/issues/36222#issuecomment-511973898 (things are going to change again a bit..).

**Comment by rynowak** [comment]
It seems like most of the overlap here is with generic-typed fields and properties right?

**Comment by roji** [comment]
I think so, but there are also some other cases. For example, you may want to have a property that can be set to null but will never return null (e.g. because it will return some sort of default). Exactly how we interpret that in ASP.NET/EF is also not immediately obvious.

**Pull request merged**